### PR TITLE
[pvr/videoplayer] fix commercial skip (EDL's) from recordings and library sources

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Fixes commercial skipping from both PVR recordings and library sources reported on 18.4.

## Motivation and Context

Broken as per this issue:
https://github.com/xbmc/xbmc/issues/16564

Since commit e68bc9f22a576bc143014a7c3b46108a0388402f frame rate and scale are no longer loaded for PVR video when required to load the EDL file. Without these the file i skipped.

This can be worked around by setting a fallback framerate in `PVR Settings->Playback`.

## How Has This Been Tested?

On OSX using latest master. Used a pvr recording with a valid EDL file and tested from both the recordings in PVR UI and as a library source. In both cases the EDL is loaded correctly.

Also tested using frame number in an EDL file and the the loading of the EDL fails as expected for a TS file.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by Rechi** [comment]
How does this PR fix commercial skip in Kodi 18.4?

**Comment by phunkyfish** [comment]
Description updated @Rechi

**Comment by Rechi** [comment]
Where, PR and commits message still mention Kodi 18.4.

**Comment by phunkyfish** [comment]
Sorry, I get you. Updating.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 1]
```diff
@@ -600,11 +600,26 @@ bool CDVDDemuxFFmpeg::Open(std::shared_ptr<CDVDInputStream> pInput, bool streami
...
+      }
+    }
+
     std::shared_ptr<CDVDInputStream> pInputStream = m_pInput;
     Dispose();
     if (!Open(pInputStream, false))
       return false;
     m_pFormatContext->duration = duration;
+    if (streamIndex >= 0)
+      m_pFormatContext->streams[streamIndex]->r_frame_rate = frameRate;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: That won't work. Video stream might not have been discovered here. I would save framerate and assign it directly to CDemuxStreamVideoFFmpeg in AddStream later.

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]: Ok, got you. Thanks.

**Comment by phunkyfish** [comment]
Ok, updated. Are there any places where `fpsRate` and `fpsScale` would be required to be set back to zero?

**Comment by FernetMenta** [comment]
After a closer look I am the opinion that this is a hack and wrong. Framerate at open is a hint (this is why the struct it is saved to is called hint) and can change later.

For the majority of EDL entries, framerate is not even required. I would drop this condition:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L3605

Then fail inside ReadEditDecisionLists() for the rare cases where framerate is required and not given directly by the EDL file. Maybe those cases don't even exist anymore.

**Comment by phunkyfish** [comment]
From `ReadEditDecisionLists()`:

```
   * The frame rate hints returned from ffmpeg for the video stream do not appear to take into
   * account whether the content is interlaced. This affects the calculation to time offsets based
   * on frames per second as most commercial detection programs use full frames, which need two
   * interlaced fields to calculate a single frame so the actual frame rate is half.
   *
   * Adjust the frame rate using the detected frame rate or height to determine typical interlaced
   * content (obtained from http://en.wikipedia.org/wiki/Frame_rate)
   *
   * Note that this is a HACK and we should be able to get the frame rate from the source sending
   * back frame markers. However, this doesn't seem possible for MythTV.
```

**Comment by phunkyfish** [comment]
So if we don't supply it we will probably break things for MythTV which I don't think we should do.

**Comment by phunkyfish** [comment]
I would lean towards fix the issue now and improve EDL handling as a separate PR as it would be a longer term change which would need testing across all PVR addons.

**Comment by FernetMenta** [comment]
> So if we don't supply it we will probably break things for MythTV

Wrong. MythTV uses comskip and comskip includes framerate in the generated comskip file. This has been changes long long ago. The hack only applies to the ancient version where framerate was not in the comsip file. I doubt that anybody has seen this message in a recent log:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/Edl.cpp#L345

**Comment by phunkyfish** [comment]
Ok, will make the required changes later and update the PR.

**Comment by phunkyfish** [comment]
Updated to completely removes the use of a supplied frame rate to EDLs. 

@emveepee can you check the solution in this PR works for your use case?

File: xbmc/cores/VideoPlayer/Edl.cpp [hunk 2]
```diff
@@ -231,7 +191,9 @@ bool CEdl::ReadEdl(const std::string& strMovie, const float fFramesPerSecond)
       }
       else if (strFields[i][0] == '#') // #12345 format for frame number
       {
-        iCutStartEnd[i] = (int64_t)(atol(strFields[i].substr(1).c_str()) / fFramesPerSecond * 1000); // frame number to ms
+        CLog::Log(LOGERROR, "%s - Frame number not supported in EDL files - supplied frame number: %s",
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/Edl.cpp]: this does not log the class name on Linux, which makes reading logs quite hard.

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 3]
```diff
@@ -3602,11 +3602,7 @@ bool CVideoPlayer::OpenVideoStream(CDVDStreamInfo& hint, bool reset)
 
     // look for any EDL files
     m_Edl.Clear();
-    if (hint.fpsrate > 0 && hint.fpsscale > 0)
-    {
-      float fFramesPerSecond = (float)m_CurrentVideo.hint.fpsrate / (float)m_CurrentVideo.hint.fpsscale;
-      m_Edl.ReadEditDecisionLists(m_item, fFramesPerSecond, m_CurrentVideo.hint.height);
-    }
+    m_Edl.ReadEditDecisionLists(m_item);
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: I wouldn't have removed fFramesPerSecond completely, just set it to zero if not supplied. Then fail in ReadEdl only if framerate required and not supplied. Containers like mkv do supply framerate correctly.

File: xbmc/cores/VideoPlayer/Edl.cpp (hunk 2 above)
Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/Edl.cpp]: Ok

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp (hunk 3 above)
Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp]: Ok

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp]: But that leaves the hack in `ReadEditDecisoonLists` around frames per second then. Why not completely clean it up?

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp]: BTW. I have another version ready of it just being set to zero. As I was doing It was thinking if these Comskip and EDL files are rare they would be rare for every format not just TS.

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp]: > But that leaves the hack in ReadEditDecisoonLists around frames per second then.

Wrong. The hack is for mpeg-ts. Containers like mkv provide correct framerate.

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp]: Ok, I get you now. Will update with the other version that just removes the hack. Thanks again for your help.

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp]: So the signature will pass the frame rate but no longer the height.

File: xbmc/cores/VideoPlayer/Edl.cpp [hunk 4]
```diff
@@ -45,48 +45,8 @@ void CEdl::Clear()
   m_lastCutTime = 0;
 }
 
-bool CEdl::ReadEditDecisionLists(const CFileItem& fileItem, const float fFrameRate, const int iHeight)
+bool CEdl::ReadEditDecisionLists(const CFileItem& fileItem, const float fFramesPerSecond)
```
Comment (ksooo) [inline xbmc/cores/VideoPlayer/Edl.cpp]: While you are at it: It does not make sense to declare by value parameters with primitve types as `const`.

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/Edl.cpp]: I did notice that but decided it was for reasons beyond my understanding 😉

Will fix it up shortly.

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/Edl.cpp]: Although, would this not signify that this value is not modified in this method?

Comment (ksooo) [reply xbmc/cores/VideoPlayer/Edl.cpp]: keep it, not a real issue, more a matter of taste. sorry for the noise.

Comment (notspiff) [reply xbmc/cores/VideoPlayer/Edl.cpp]: indeed, see discussion in slack.

Comment (phunkyfish) [reply xbmc/cores/VideoPlayer/Edl.cpp]: NP

**Comment by phunkyfish** [comment]
Ok, should be good now, any other notes @FernetMenta ?

**Comment by Rechi** [comment]
jenkins build this please

**Comment by emveepee** [comment]
> @emveepee can you check the solution in this PR works for your use case?

Sorry I have no way of building matrix at this point and the binary isn't available in the artifacts.  If I was testing this I would change the edl to use frames rather then times similar to the .txt file and use # to identify it is sent as frame not as time.

**Comment by phunkyfish** [comment]
I can create a WIP PR for Leia for you to test with tomorrow.

**Comment by phunkyfish** [comment]
Unless there are any objections I will merge this tomorrow.

**Comment by bugwelle** [comment]
Just want to give my two cents. Making the frame rate in Comskip files mandatory is fine, we just need to remember to update https://kodi.wiki/view/Edit_decision_list#Comskip when this is merged as not all examples there use the frame rate.

**Comment by phunkyfish** [comment]
It would only be mandatory for TS files. But yes, I can update the Wiki for both EDL and Comskip files.

**Pull request merged**